File: src/main/python/AoC2024_23.py
#! /usr/bin/env python3
#
# Advent of Code 2024 Day 23
#

import sys
from collections import defaultdict

from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.common import aoc_samples

Input = dict[str, set[str]]
Output1 = int
Output2 = str


TEST = """\
kh-tc
qp-kh
de-cg
ka-co
yn-aq
qp-ub
cg-tb
vc-aq
tb-ka
wh-tc
yn-cg
kh-ub
ta-co
de-co
tc-td
tb-wq
wh-td
ta-ka
td-qp
aq-cg
wq-ub
ub-vc
de-ta
wq-aq
wq-vc
wh-yn
ka-de
kh-ta
co-tc
wh-qp
tb-vc
td-yn
"""


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        edges = defaultdict[str, set[str]](set)
        for line in input_data:
            node_1, node_2 = line.split("-")
            edges[node_1].add(node_2)
            edges[node_2].add(node_1)
        return edges

    def part_1(self, edges: Input) -> Output1:
        return len(
            {
                tuple(sorted((t, neighbour_1, neighbour_2)))
                for t in (comp for comp in edges if comp.startswith("t"))
                for neighbour_1 in edges[t]
                for neighbour_2 in edges[neighbour_1]
                if t in edges[neighbour_2]
            }
        )

    def part_2(self, edges: Input) -> Output2:
        cliques = [{comp} for comp in edges]
        for clique in cliques:
            for comp in edges:
                if clique & edges[comp] == clique:
                    clique.add(comp)
        return ",".join(sorted(max(cliques, key=len)))  # type:ignore[arg-type]

    @aoc_samples(
        (
            ("part_1", TEST, 7),
            ("part_2", TEST, "co,de,ka,ta"),
        )
    )
    def samples(self) -> None:
        pass


solution = Solution(2024, 23)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()

File: src/main/java/AoC2024_23.java
import static java.util.Comparator.comparing;
import static java.util.stream.Collectors.joining;

import com.github.pareronia.aoc.IterTools.ProductPair;
import com.github.pareronia.aoc.StringOps;
import com.github.pareronia.aoc.StringOps.StringSplit;
import com.github.pareronia.aoc.solution.Sample;
import com.github.pareronia.aoc.solution.Samples;
import com.github.pareronia.aoc.solution.SolutionBase;

import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

public final class AoC2024_23 extends SolutionBase<Map<String, Set<String>>, Integer, String> {

    private AoC2024_23(final boolean debug) {
        super(debug);
    }

    public static AoC2024_23 create() {
        return new AoC2024_23(false);
    }

    public static AoC2024_23 createDebug() {
        return new AoC2024_23(true);
    }

    @Override
    protected Map<String, Set<String>> parseInput(final List<String> inputs) {
        final Map<String, Set<String>> edges = new HashMap<>();
        for (final String line : inputs) {
            final StringSplit sp = StringOps.splitOnce(line, "-");
            edges.computeIfAbsent(sp.left(), k -> new HashSet<>()).add(sp.right());
            edges.computeIfAbsent(sp.right(), k -> new HashSet<>()).add(sp.left());
        }
        return edges;
    }

    @Override
    public Integer solvePart1(final Map<String, Set<String>> edges) {
        record Triangle(String v1, String v2, String v3) {
            public Triangle sorted() {
                final String[] v = {v1, v2, v3};
                Arrays.sort(v);
                return new Triangle(v[0], v[1], v[2]);
            }
        }
        return (int)
                edges.keySet().stream()
                        .filter(comp -> comp.startsWith("t"))
                        .flatMap(t -> edges.get(t).stream().map(n1 -> new ProductPair<>(t, n1)))
                        .flatMap(
                                pp ->
                                        edges.get(pp.second()).stream()
                                                .filter(n2 -> edges.get(n2).contains(pp.first()))
                                                .map(
                                                        n2 ->
                                                                new Triangle(
                                                                        pp.first(),
                                                                        pp.second(),
                                                                        n2)))
                        .map(Triangle::sorted)
                        .distinct()
                        .count();
    }

    @Override
    public String solvePart2(final Map<String, Set<String>> edges) {
        final List<HashSet<String>> cliques =
                edges.keySet().stream().map(comp -> new HashSet<>(Set.of(comp))).toList();
        cliques.forEach(
                clique ->
                        edges.entrySet().stream()
                                .filter(e -> e.getValue().containsAll(clique))
                                .forEach(e -> clique.add(e.getKey())));
        final Set<String> largest =
                cliques.stream().max(comparing(HashSet<String>::size)).orElseThrow();
        return largest.stream().sorted().collect(joining(","));
    }

    @Override
    @Samples({
        @Sample(method = "part1", input = TEST, expected = "7"),
        @Sample(method = "part2", input = TEST, expected = "co,de,ka,ta")
    })
    public void samples() {}

    public static void main(final String[] args) throws Exception {
        AoC2024_23.create().run();
    }

    private static final String TEST =
            """
            kh-tc
            qp-kh
            de-cg
            ka-co
            yn-aq
            qp-ub
            cg-tb
            vc-aq
            tb-ka
            wh-tc
            yn-cg
            kh-ub
            ta-co
            de-co
            tc-td
            tb-wq
            wh-td
            ta-ka
            td-qp
            aq-cg
            wq-ub
            ub-vc
            de-ta
            wq-aq
            wq-vc
            wh-yn
            ka-de
            kh-ta
            co-tc
            wh-qp
            tb-vc
            td-yn
            """;
}
